### ISSUE TITLE
Proposal: Support for iterating registered names/introspection

### ISSUE THREAD

**Issue by emilong** [young-steveo/bottlejs#51]
Hi all,

I've been exploring a pattern in using bottlejs where I namespace a certain set of similar objects (in my case tasks) and then I iterate them for registration. It might look something like:

```javascript
bottle.service('task.Foo', function() {...})
bottle.service('task.Bar', function() {...})
```

Then using (again, as an example) [kue](https://github.com/Automattic/kue):

```javascript
const kue = require('kue')
const queue = kue.createQueue()
Object.values(bottle.container.task).forEach((task) => {
  if (!task.type) return
  queue.process(task.type, task.run)
})
```

Now, this works at the moment and for this example because I'm specially tagging the tasks with a "task" key, but that pattern may not be appropriate for other cases and definitely feels a bit hacky.

What do other folks think about adding a supported API for iterating and/or introspecting registered services, factories, or whatever has a name in a bottle? (Or does this exist and I missed it?)

Thanks!

**Comment by young-steveo** [comment]
You can use the `digest` method to instantiate and return a list of services:

https://jsfiddle.net/kvrjr1ze/

So, your final code would look like:

```js
// assuming Foo and Bar are not namespaced.

const kue = require('kue');
const queue = kue.createQueue();
bottle.digest(['Foo', 'Bar']).forEach(task => {
  if (!task.type) return
  queue.process(task.type, task.run)
});
```

However, I wouldn't call your original method hack-y; as you can see, the bottle.digest method needs to be told what to digest, and by registering your Task services under a namespace, you don't have to manage a separate list (in the example case, the array passed to digest is hard coded).

Another option would be to combine the two techniques:


```js
// assuming Foo and Bar are namespaced with 'task.'

const kue = require('kue');
const queue = kue.createQueue();
const tasks = Object.keys(bottle.container.task).map(key => 'task.' + key);
bottle.digest(tasks).forEach(task => {
  if (!task.type) return
  queue.process(task.type, task.run)
});
```

**Comment by emilong** [comment]
Ok, cool. I was thinking it was hacky because if you inspect the bottle, it has the providers and other keys which seem like implementation details. Rather than sorting out between those, I thought it might be better to provide an explicit API, but if this approach will be stable, I'll just keep doing it. 😄

**Comment by young-steveo** [comment]
Actually, now that you mention it you do end up having to filter out bottle methods like `$register`.  So maybe it would be worthwhile to add something to the api that just gives you the service keys.

**Comment by young-steveo** [comment]
@emilong I've got a preliminary implementation working in my local build:

```js
// all of these will filter out the $register and $list methods and the Providers,
// effectively returning a list of service strings

var bottle = new Bottle();

bottle.service('A', function() {});

Bottle.list(bottle.container); // static method on Bottle
bottle.list(); // instance method looks at it's own container
bottle.container.$list(); // container method looks at itsself

**Comment by young-steveo** [comment]
I'm going to test it out some more and it'll land in bottle `v1.3`

**Comment by emilong** [comment]
Awesome! Then to reproduce the example above, something like this?

```javascript
const kue = require('kue');
const queue = kue.createQueue();
const tasks = bottle.container.tasks.$list();
bottle.digest(tasks).forEach(task => {
  if (!task.type) return
  queue.process(task.type, task.run)
});
```

Just curious, why prefix with a '$' on the container version?

Thanks!

**Comment by young-steveo** [comment]
@emilong If all of your services under the `task` namespace have a `type` property, you don't need to return early in the foreach either, but yes, that's essentially what you'll do.

I'm using the `$` prefix on the container version in case someone wants to register a service called `list` for themselves.  This convention is already in place for `bottle.container.$register` which is a copy of `bottle.register` that exists on the container.

**Comment by emilong** [comment]
Whoops, a bit of copy pasta there... Thanks for the explanation on the $ prefix! Can't wait to use this! :)

**Comment by young-steveo** [comment]
Now available in [v1.3.0](https://github.com/young-steveo/bottlejs/releases/tag/v1.3.0)
